### DISCUSSION TITLE
Are keywords supported and if so where are they stored?

### DISCUSSION THREAD

**Discussion by bverkron** [immich-app/immich#13450] [Q&A] [answered]
I see some discussion about keywords here and there but I can't get a clear understanding if they're supported, where to find them, if they are used in searches, etc.

For example PR https://github.com/immich-app/immich/pull/11980 mentions reading keywords from EXIF but I can't seem to determine if they should show up in the image info in Immich or be searchable.

Furthermore I'm specifically trying to get Immich to read keywords that Lightroom has included in files that I've exported but it seems those are stored in the IPTC metadata and not the EXIF metadate. Does this mean Immich cannot read those keywords?

**Comment by mertalev** [comment, marked as answer]
You have to enable tags in the user settings. It's a bit buried. But they should be ready to view from the sidebar after that without needing to process anything again.

**Reply by bverkron** [reply]
Thanks! Found it under User icon (top-right) > Account Settings > Features > Tags

Doesn't seem to be anything in the doc about this currently. Perhaps it's behind as this is a new-ish feature?

Either way, thank you. I am EXTREMELY impressed by Immich so far. It lives up to the hype!! 💙

**Comment by ok1ok1** [comment]
I have tags enabled, and sure enough, all of my keywords are showing up as tags. There’s a huge problem though, the way Immich handles this data is practically useless. Nearly every other photo management system makes keywords searchable by default. You just type a keyword or even multiple keywords and any photo containing them is returned. Google Photos, Amazon Photos, Synology Photos, SmugMug, and others all work this way. I’m honestly shocked that Immich didn’t think to include a dedicated keyword category.

Unless I’m missing something, Immich requires you to manually select search options and choose a specific keyword, now tag, from a dropdown menu. For images that contain multiple keywords, the entire string is treated as a single unique entry.

Here’s a quick example of why this is so limiting:

Joseph, 4th of July, Katie

Joseph

Katie

In Immich, I can only filter by one of these entries, and only exact matches are returned. I can’t simply search for 4th of July, Katie — that won’t return anything. Likewise, if I search for Joseph, only photos tagged with exactly that single keyword will appear. Immich’s lack of foresight here essentially destroys all the effort that went into properly keywording these photos in the first place. This limitation is the only thing preventing me from using Immich as my primary photo management tool.

It’s possible that I’m missing something and there’s a feature I’m unaware of, but as it stands, either Immich is shortsighted and poorly designed, or I’m completely misunderstanding how it’s supposed to work. Honestly, I hope it’s me.

**Reply by mertalev** [reply]
There's no requirement to search by only one tag; you should be able to add both 4th of July and Katie through the dropdown.

**Reply by ok1ok1** [reply]
If I have those individual keywords.... Even then it would only match Katie or 4th of July. Not an image with keywords for both.  

When immich it converts keywords to tags it treats all of the keywords as a single tag. It doesn't separate them out. Even if I included both individual tags Katie & 4th of July it would not match a photo tagged "Katie, 4th of July".

See this mess? These are my keywords converted to tags.... It does not split them into individual tags as expected. Beyond that the drop down implementation for keywords is just silly. It's not a good solution. Keywords are not tags and they need their own implementation. Keywords are a long standing feature of most common image formats. How this was so utterly overlooked is beyond me. 
 
<img width="457" height="132" alt="image" src="https://github.com/user-attachments/assets/37822148-44b9-40e1-b475-36b53a5aa710" />

I'd need to type the entire strings exactly for them to be retrieved. Immich cannot retrieve a partial match.

**Reply by bverkron** [reply]
> If I have those individual keywords.... Even then it would only match Katie or 4th of July. Not an image with keywords for both.

This seems to be the case for me as well, at least with the main default search field. However, if I open the advanced search options search for multiple partial keywords (space separated) it works as expected and finds photos with either keyword partially matched. To use your example if I searched for "katie 4th" it should return photos with either. Would be nicer if this also worked in the main search filed by default.

<img width="147" height="91" alt="image" src="https://github.com/user-attachments/assets/1f8f1db1-5416-47a1-9ece-db60cad296de" />

Edit: Part of the issue with the default search might be that it defaults to Filename, at least in the web UI.
<img width="617" height="67" alt="Screenshot 2025-10-26 at 12 11 36 PM" src="https://github.com/user-attachments/assets/f9560ce8-9c13-4ca7-8970-88d143b9ee77" />

> See this mess? These are my keywords converted to tags.... It does not split them into individual tags as expected.

What is your source for this data? It looks like the keywords are wedged together with no space between the comma and the next keyword. That might have something to do with it. I see a single keyword in my Immich instance that is two wedged together with a comma in between an no spaces like that, though all my other ones are fine. There might be room for improvement here when reading the source data.

You could use exiftool or something on one of the original images you uploaded to Immich and see how it's storing the keyword data. That might give a clue as to why Immich is treating it as a single wedged together keyword.

**Reply by ok1ok1** [reply]
I was wrong to assume everyone was experiencing this issue. It looks like the problem on my end was caused by an earlier version of Immich. I’ve noticed that new imports now split keywords correctly. The good news is it doesn’t seem to be an ongoing Immich issue. The bad news is I’ve got a big cleanup ahead of me.

I’ve written a couple of scripts that use the API to help with the cleanup. The first one finds all tags containing commas, retrieves the associated images, splits each comma-separated value into individual tags, and applies them. It then removes the original combined tag from those images.

A follow-up script deletes any tags that no longer have assets attached.

I still have a bit of cleanup left, but I’m most of the way there.  — Keywords are finally usable.
